util/file_reader_writer.cc

```diff
@@ -512,19 +513,24 @@ class ReadaheadRandomAccessFile : public RandomAccessFile {
         return Status::OK();
       }
     }
+    size_t aligned_offset = TruncateToPageBoundary(alignment_, offset + copied);
     Slice readahead_result;
-    Status s = file_->Read(offset + copied, readahead_size_, &readahead_result,
-      buffer_.get());
+    Status s = file_->Read(aligned_offset, readahead_size_ + alignment_,
+                           &readahead_result, buffer_.get());
     if (!s.ok()) {
       return s;
     }
-
-    auto left_to_copy = std::min(readahead_result.size(), n - copied);
-    memcpy(scratch + copied, readahead_result.data(), left_to_copy);
-    *result = Slice(scratch, copied + left_to_copy);
+    size_t offset_advance = offset - aligned_offset;
```

Comment:
Yes, I see your issue. It should be `size_t offset_advance = offset + copied - aligned_offset;`